src/DurableTask.Netherite/StorageLayer/Faster/StoreWorker.cs

```diff
@@ -349,7 +394,120 @@ void ScheduleNextIdleCheckpointTime()
+                            if (token.HasValue)
+                            {
+                                this.traceHelper.FasterProgress("Checkpointing state machine: store checkpoint started");
+                                this.pendingIndexCheckpoint = null;
+                                this.pendingStoreCheckpoint = this.WaitForCheckpointAsync(false, token.Value, true);
+                                this.numberEventsSinceLastCheckpoint = 0;
+                            }
+                            var checkpointStarted = token.HasValue;
+                            return checkpointStarted;
+
+                        },
+                        messageOnError: "Could not start store checkpoint before timeout");
+                }
+            }
+            else if (this.pendingCompaction != null)
+            {
+                if (this.pendingCompaction.IsCompleted == true)
+                {
+                    this.traceHelper.FasterProgress("Checkpointing state machine: pendingCompaction has completed");
+                    await this.pendingCompaction; // observe exceptions here
+
+                    // force collection of memory used during compaction
+                    GC.Collect();
+
+                    // the index checkpoint is next
+                    this.StartCheckpointOrFailOnTimeout(
+                        checkpointRoutine: () =>
+                        {
+                            var token = this.store.StartIndexCheckpoint();
+                            if (token.HasValue)
+                            {
+                                this.traceHelper.FasterProgress("Checkpointing state machine: index checkpoint started");
+                                this.pendingCompaction = null;
+                                this.pendingIndexCheckpoint = this.WaitForCheckpointAsync(true, token.Value, false);
+                            }
+                            var checkpointStarted = token.HasValue;
+                            return checkpointStarted;
+                        },
+                        messageOnError: "Could not start index checkpoint before timeout");
+                }
+            }
+            else if (this.CheckpointDue(out var trigger, out long? compactUntil))
+            {
+                this.traceHelper.FasterProgress($"Checkpointing state machine: checkpoint is due. Trigger='{trigger}'. compactUntil='{compactUntil}'");
+                this.pendingCheckpointTrigger = trigger;
+
+                var isCompactionSkipped = !compactUntil.HasValue;
+                this.pendingCompaction = isCompactionSkipped ? Task.FromResult((long?)null) : this.RunCompactionAsync(compactUntil.Value);
+
+                if (isCompactionSkipped)
+                {
+                    this.traceHelper.FasterProgress($"RunCompactionAsync was skipped because a compaction range was not provided by trigger = {trigger}.");
+                }
```

Comment:
@cgillum: the log informing the user that compaction was skipped was moved here btw. In this position, we're able to clarify that the reason why we skipped compaction is because a given compaction trigger did not specify a compaction range.